### ISSUE TITLE
Wrong behaviour on Screen rotation or Windows resize

### ISSUE THREAD

**Issue by ChibuikeOruche** [navit-gps/navit#433]
Navit app is a free mobile application which offers an offline navigation system on your mobile device.

#### Expected behavior
When I click on a menu item (or option and rotate the screen, I should remain on my current view and continue with whatever I intended to do with the app.

#### Actual behavior
Unlike the expected behavior, rotating my device after clicking on a submenu reverts me back to the menu thus making me start all over with what I want to do in the app. *It occurs on all submenu items*. 

#### How to reproduce
Open the Navit app. Click on the menu button(bottom right of screen). A menu will be displayed. Click on any menu option of choice. It should display the options in that submenu. Then, rotate your screen.

* Operating system: Android 4.4.2(KitKat)
* Device: Tecno Y6 Build version:1.7.1

#### Recording Of The Bug
![Navit_bug.mp4](https://cdn.utopian.io/posts/3a492cebe4ebca72bc11fdcc2bda70735fdaNavit_bug.mp4)

    

<br /><hr/><em>Posted on <a href="https://utopian.io/utopian-io/@goodluckleo/wrong-behaviour-on-screen-rotation">Utopian.io -  Rewarding Open Source Contributors</a></em><hr/>

**Comment by metalstrolch** [comment]
Same happens on resizing Navit on desktop.

So not directly related to rotation nor related to a special graphics backend / OS build. I think it's only related to resizing, as rotation usually imposes resizing the screen as well.

**Comment by lains** [comment]
This forced return to the main menu is indeed quite annoying on the current internal GUI.
This can happen frequently while manipulating the device, and while entering GPS coords or selecting a street in a town, this makes the user loose time going again through the GUI a second time.

As far a code is concerned, this behaviour comes from function `gui_internal_resize()` in https://github.com/navit-gps/navit/blob/trunk/navit/gui/internal/gui_internal.c
In that function, when resizing the screen (or rotating a portable device), a variable named `changed` will be set to 1, which later on will make `gui_internal_html_main_menu(this);` be called. This is what forces the return to the main menu, and from the code, it seems it is the desired behaviour.
This has the effect of cancelling the current menu (prune) and reload the main menu.

Fixing this is not trivial though...

The behaviour can be slightly improved by reloading the current menu (submenu or main menu) using the following code instead of calling `gui_internal_html_main_menu()`:
```
        gui_internal_prune_menu(this, NULL);
        char *href=g_strdup(this->href);
        gui_internal_html_load_href(this, href, 0);
        g_free(href);
```

But this will only work for menus/submenus, not for any currently displayed command/form GUI. We would return to the last submenu.
Unfortunately, calling `gui_internal_menu_render(this);` instead fixes the issue but but will not take the new display/window size into account.
The GUI will thus still be fitted to the initial display size, and blank space will be displayed to fill-in the additional space on the right and bottom (or GUI will be cropped on the right and bottom sides).

Probably, the real solution to this is to allow widgets to recompute their appearance given a new size (width and height provided), this means adding new resize methods for all widgets and probably repack them all (if we have enough internal data to do this) each time we resize.
This could mean quite some code rework...

**Comment by lains** [comment]
This seems to be a lot of work from what I could test.
Vertical lists are natively displaying properly when resizing vertically (increase or decrease of space), however they do not expand or reduct properly when resizing horizontally.
The first solution would be to reset packing of all content of currently displayed screen, but even if there is a `gui_internal_widget_reset_pack()` function, and then resetting width and height to the new screen size, it seems just calling again `gui_internal_widget_pack()` and render() on the root widget does not handle a resize properly...
I think this would need to do a rewrite of all widgets's pack() functions in order to be able to run twice on a widget.

**Comment by lains** [comment]
Trying to submit a solution for this in PR #682.
That PR is partially solving this for now... it still needs more coding.

**Comment by lains** [comment]
I'm pretty sure I will get it to work as expected on the internal GUI.
Just bear with me a few more weeks, I'm working on another PR...

**Comment by lains** [comment]
I'm almost there but the code around widget packing is quite complex. I still need to perform some cleanup and I will push new code.
However, I am getting small deltas between an initially small rendered window resized to fullscreen, and an originally fullscreen rendered one...
Is there any internal GUI expert that could help out?

**Comment by lains** [comment]
This is now fixed since commit 94fccd71c7228ec2de548ae35d5d3cd12c6a100a
